library/Opus/Model2/Language.php

```diff
@@ -0,0 +1,395 @@
+ *
+ * LICENCE
+ * OPUS is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU General Public License as published by the Free Software
+ * Foundation; either version 2 of the Licence, or any later version.
+ * OPUS is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details. You should have received a copy of the GNU General Public License
+ * along with OPUS; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ * @copyright   Copyright (c) 2021, OPUS 4 development team
+ * @license     http://www.gnu.org/licenses/gpl.html General Public License
+ *
+ * @category    Framework
+ * @package     Opus\Db2
+ * @author      Jens Schwidder <schwidder@zib.de>
+ */
+
+namespace Opus\Model2;
+
+use Doctrine\ORM\Mapping as ORM;
+use Doctrine\ORM\ORMException;
+
+/**
+ * @ORM\Entity(repositoryClass="Opus\Db2\LanguageRepository")
+ * @ORM\Table(name="languages")
+ */
+class Language extends AbstractModel
+{
+    public const PROPERTY_COMMENT = 'Comment';
+    public const PROPERTY_PART2B  = 'Part2B';
+    public const PROPERTY_PART2T  = 'Part2T';
+    public const PROPERTY_PART1   = 'Part1';
+    public const PROPERTY_SCOPE   = 'Scope';
+    public const PROPERTY_TYPE    = 'Type';
+    public const PROPERTY_REFNAME = 'RefName';
+    public const PROPERTY_ACTIVE  = 'Active';
+
+    /**
+     * @ORM\Id
+     * @ORM\Column(type="integer")
+     * @ORM\GeneratedValue
+     *
+     * @var int
+     */
+    private $id;
+
+    /**
+     * @ORM\Column(type="string", name="part2_b", length=3, options={"fixed" = true})
+     *
+     * @var string
+     */
+    private $part2B;
+
+    /**
+     * @ORM\Column(type="string", name="part2_t", length=3, options={"fixed" = true})
+     *
+     * @var string
+     */
+    private $part2T;
+
+    /**
+     * @ORM\Column(type="string", length=2, options={"fixed" = true})
+     *
+     * @var string
+     */
+    private $part1;
+
+    /**
+     * @ORM\Column(type="string", columnDefinition="ENUM('I','M','S')")
+     *
+     * @var string
+     */
+    private $scope;
+
+    /**
+     * @ORM\Column(type="string", columnDefinition="ENUM('A','C','E','H','L','S')")
+     *
+     * @var string
+     */
+    private $type;
+
+    /**
+     * @ORM\Column(type="string", name="ref_name", length=150)
+     *
+     * @var string
+     */
+    private $refName;
+
+    /**
+     * @ORM\Column(type="string", length=150)
+     *
+     * @var string
+     */
+    private $comment;
+
+    /**
+     * @ORM\Column(type="smallint")
+     *
+     * @var int
+     */
+    private $active = 0;
+
+    /**
+     * @return int
+     */
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param int $id
+     */
+    public function setId($id)
```

Comment:
Wird mit Doctrine normalerweise eine setId-Funktion verwendet? In der Dokumentation habe ich nichts konkretes gefunden. Ich glaube manipuliert die Variable direkt über Reflection. Der Code soll hier erst einmal nicht geändert werden, aber ich denke das ist ein offenes Thema. Im Prinzip sollte die ID ein ReadOnly-Property sein bzw. Set-Once, aber beim Import der Master-Daten nach der Installation, wäre es nützlich, die automatische ID-Generierung abzuschalten und die IDs explizit zu setzen. Das sollte mit Doctrine möglich sein. Dafür gibt es separate Issues.

Reply:
Die Tests laufen auch durch, wenn die `setId`-Funktion nicht existiert. Ich würde die Funktion erst einmal weglassen. Doctrine verwendet Reflection to set the load the values. Das ID-Property sollte auch in der Basis-Klasse umsetzbar sein, oder?